JBTM-2812 Ensure JTS/JTA forget behaviour is consistent

mmusgrov opened the pull request:
https://issues.jboss.org/browse/JBTM-2812
!BLACKTIE !XTS !QA_JTA !QA_JTS_JACORB !QA_JTS_JDKORB !PERF

tomjenkinson commented on ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java:
```diff
@@ -909,14 +909,15 @@ public void forget() throws org.omg.CORBA.SystemException
 
-		handleForget() ;
+		if (!handleForget())
+		    throw new UNKNOWN(); // will cause log to be rewritten.
```
Great find - not sure how I missed that one!

tomjenkinson commented:
Marking as ready to merge but I can see the BT tests haven't completed yet so it would be contingent on that.

mmusgrov commented:
@tomjenkinson I do have a couple of more changes to push. I will put those in a separate commit and squash if you are ok with them.

mmusgrov commented:
@tomjenkinson I added a small extra commit

tomjenkinson commented on ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java:
```diff
@@ -909,14 +909,15 @@ public void forget() throws org.omg.CORBA.SystemException
+	private boolean handleForget()
 	{
 		if ((_theXAResource != null) && (_tranID != null))
 		{
```
I think you need to move the _heuristic = TwoPhaseOutcome.FINISH_OK; to the same place as the local JTA one.

tomjenkinson commented on ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java:
```diff
@@ -928,8 +929,11 @@ private void handleForget()
 			}
 			catch (Exception e)
 			{
+                return false;
```
Could add the warning as you did in local JTA

tomjenkinson commented:
I am OK with the changes but I think you need to reflect them in the JTS (log message and _heuristic setting).

Also the JBTM- issue needs some rewording :)

tomjenkinson commented:
Still OK with a merge if you make the changes as discussed above. Will revert my flag for now though so its clearer.

mmusgrov commented:
The issue title is: "JBTM-2812 Ensure failed forget calls are reported"
I am also reported the underlying cause (e.getMessage()) in the UNKNOW exception.
I am issuing a log warning if forget fails in the JTS case (just like I was in the JTA case).

What did you mean by your other point: "reflect them in the JTS _heuristic setting"?

tomjenkinson commented:
The issue title is not what you quoted above

tomjenkinson commented on ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java:
```diff
@@ -909,14 +915,14 @@ public void forget() throws org.omg.CORBA.SystemException
         }
 
-		handleForget() ;
+		handleForget();
```
superfluous?

mmusgrov replied:
I meant I squashed the commits and reworded the title.

tomjenkinson commented:
Regarding _heuristic:

1. Look at https://github.com/jbosstm/narayana/pull/1100/files
2. Look at how you moved _heuristic for the arjunacore/XARR
3. You didn't make the same change to the orbspecific/XARR - looking at the file it looks susceptible to the same problem you likely moved the _heuristic setting for in ac/XARR

mmusgrov commented:
I changed the issue title and squashed the commits in response to your comment (the PR comment will not reflect that).

Good catch regarding not updating the heuristic in the failure case.

tomjenkinson commented:
Thanks Mike.

I just want to check something because I am not following 100%. I understood that we want to retain the log now? If that is the case the description and title in https://issues.jboss.org/browse/JBTM-2812 need an update as you are talking about "reverting" the retention of the log in the desc and "always remove" in the title.

tomjenkinson commented:
I should say, patch LGTM (if it does what I think it does, and that is retain the log if the forget fails.

mmusgrov commented:
The issue title is currently: "JBTM-2812 Ensure failed forget calls are reported"
How about:
   "JBTM-2812 Ensure we retain a log if the resource forget call fails"

tomjenkinson commented:
Wonderfully crafted and poetic - many thanks!

mmusgrov commented:
I will amend the commit message to reflect the new well crafted title: "Retain a log if a forget call on a resource fails" ;)

And I think that the new jira description is equally poetic too:

> When we clear a heuristic from XAResourceRecord we call forget on the resource and then delete the record even if the forget operation fails. The preferred behaviour is to retain the record if the forget fails so that it can be reported via our tooling and so that we still have the option of later retrying the forget operation (also via the tooling).

tomjenkinson commented:
Indeed, it's a heart-warming tale of triumph over adversity *****!

tomjenkinson commented:
@mmusgrov hold?

mmusgrov commented:
@tomjenkinson The hold is pending unit tests for the JTS version of XAResourceRecordMBean

tomjenkinson commented:
Ah - good catch. I have removed my RfM until I review those. Thanks

tomjenkinson commented:
ObjStoreBrowserTest  is ran and the WFLY-7785 is unrelated

mmusgrov commented:
What do you think of my new tests for the JTS variant of XAResourceRecord?

tomjenkinson commented:
If it is ObjStoreBrowserTest.java for these new XA tests then they look good

mmusgrov commented:
Sorry, I added and then immediately deleted a comment about my new tests. I deleted it since I noticed that you had already reviewed the new ObjStoreBrowserTest.java
